src/operator/deinput.rs

```diff
@@ -0,0 +1,703 @@
+            Runtime::init_circuit(workers, |circuit| {
+                let (zset, zset_handle) = circuit.add_input_zset::<TestStruct, isize>();
+                let (set, set_handle) = circuit.add_input_set::<TestStruct, isize>();
+                let (map, map_handle) = circuit.add_input_map::<i64, TestStruct, isize>();
+
+                let zset_output = zset.output();
+                let set_output = set.output();
+                let map_output = map.output();
+
+                (
+                    (zset_handle, zset_output),
+                    (set_handle, set_output),
+                    (map_handle, map_output),
+                )
+            })
+            .unwrap();
+
+        let de_zset = DeZSetHandle::new(zset_input);
+        let de_set = DeSetHandle::new(set_input);
+        let de_map = DeMapHandle::new(map_input, |test_struct: &TestStruct| test_struct.id);
+
+        (
+            dbsp,
+            (Box::new(de_zset), Box::new(de_set), Box::new(de_map)),
+            (zset_output, set_output, map_output),
+        )
+    }
+
+    // Feed `inputs` in CSV format.
+    fn insert_csv(
+        dbsp: &mut DBSPHandle,
+        input_handles: &mut InputHandles,
+        output_handles: &OutputHandles,
+        inputs: &[TestStruct],
+    ) {
+        let zset_handle = &mut input_handles.0;
+        let set_handle = &mut input_handles.1;
+        let map_handle = &mut input_handles.2;
+
+        let zset_output = &output_handles.0;
+        let set_output = &output_handles.1;
+        let map_output = &output_handles.2;
+
+        let zset = OrdZSet::from_tuples(
+            (),
+            inputs.iter().map(|v| (v.clone(), 1)).collect::<Vec<_>>(),
+        );
+        let map = <OrdIndexedZSet<i64, TestStruct, isize, usize>>::from_tuples(
+            (),
+            inputs
+                .iter()
+                .map(|v| ((v.id, v.clone()), 1isize))
+                .collect::<Vec<_>>(),
+        );
+
+        zset_handle.reserve(inputs.len());
+        set_handle.reserve(1);
+        map_handle.reserve(0);
+
+        // Serialize `inputs` as CSV.
+        let mut csv_bytes: Vec<u8> = vec![];
+        let mut csv_writer = CsvWriter::from_writer(&mut csv_bytes);
+        for input in inputs.iter() {
+            csv_writer.serialize(input).unwrap();
+        }
+        let _ = csv_writer.into_inner().unwrap();
+        let csv_string = String::from_utf8(csv_bytes).unwrap();
+
+        // println!("csv:\n{}", csv_string);
+
+        // CSV reader iterates over CSV records.
+        let mut csv_reader = CsvReader::from_reader(csv_string.as_bytes());
+
+        for input in csv_reader.records() {
+            let record = input.unwrap();
+            // Wrap CSV record in erased deserializer.
+            let mut deserializer = string_record_deserializer(&record, None);
+            let mut deserializer = <dyn ErasedDeserializer>::erase(&mut deserializer);
```

Comment:
this is odd, to create a new deserializer for each input record.

Reply:
A deserializer usually just store a reference to the underlying data, so it's not an expensive abstraction if that's your concern.

Having said that, this may not be the best API design.  It's tricky to come up with an interface that allows switching between different serialized representations of the same type at runtime.  `erased_serde` makes this possible by encapsulating the knowledge of the format in the `ErasedDeserializer` trait.  I can't think of another way.  Let's see if this even works (i.e., can we have an end-to-end data pipeline going).